File: functions/src/lib/quota.ts
// MIRROR: lib/core/constants/quota.dart exports `kQuotaTimezone = 'Asia/Dhaka'`
// (plan 03-10). If this constant drifts from the Dart side, day-key mismatch
// causes the daily quota to reset at UTC midnight instead of Dhaka midnight
// (PITFALLS #3). The shared constant is the contract.
//
// NEVER use: new Date().toISOString().slice(0, 10)  ← UTC, NOT Dhaka.
// ALWAYS use: Intl.DateTimeFormat('en-CA', { timeZone: QUOTA_TZ })  ← Dhaka.

export const QUOTA_TZ = 'Asia/Dhaka';

/**
 * Returns the Dhaka calendar day key for [now], formatted 'YYYY-MM-DD'.
 * Used as the document ID for /users/{uid}/usage/{dayKey}.
 *
 * @example
 *   // UTC 18:00 on 2026-05-18 is Dhaka 00:00 on 2026-05-19 (UTC+6)
 *   getDhakaDateKey(new Date('2026-05-18T18:00:00.000Z'))  // → '2026-05-19'
 */
export function getDhakaDateKey(now: Date = new Date()): string {
  return new Intl.DateTimeFormat('en-CA', {
    timeZone: QUOTA_TZ,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(now);
}

/**
 * Returns the Dhaka calendar month key for [now], formatted 'YYYY-MM'.
 * Used as the document ID for /system/quota/{monthKey} (D-10).
 */
export function monthKey(now: Date = new Date()): string {
  return new Intl.DateTimeFormat('en-CA', {
    timeZone: QUOTA_TZ,
    year: 'numeric',
    month: '2-digit',
  })
    .format(now)
    .slice(0, 7);
}

File: lib/core/constants/quota.dart
// MIRROR: functions/src/lib/quota.ts exports `QUOTA_TZ = 'Asia/Dhaka'` (plan 03-02).
// These two constants MUST stay in sync. If they diverge, quota day-keys differ
// between client and server — a student's display quota count drifts.
//
// Used by ChatViewModel to display remaining messages (display-only); actual
// enforcement is server-side in functions/src/lib/rate_limit.ts.
//
// NEVER use: DateTime.now().toIso8601String().substring(0, 10)  ← UTC, NOT Dhaka.
// ALWAYS use: dhakaDateKey(now) below.

import 'package:intl/intl.dart';

/// The shared quota timezone — must match `functions/src/lib/quota.ts QUOTA_TZ`.
const String kQuotaTimezone = 'Asia/Dhaka';

/// Returns the Dhaka calendar day key for [now], formatted 'YYYY-MM-DD'.
///
/// Dhaka (Bangladesh Standard Time, BST) is UTC+6 year-round; Bangladesh does
/// not observe DST. We shift [now] forward by 6 hours and format as ISO date.
///
/// Used as the read-side mirror of functions/src/lib/rate_limit.ts's
/// `getDhakaDateKey()` — ChatViewModel reads `/users/{uid}/usage/{dhakaDateKey(now)}`
/// for the display quota counter.
///
/// Example:
///   // UTC 18:00 on 2026-05-18 is Dhaka 00:00 on 2026-05-19 (UTC+6).
///   dhakaDateKey(DateTime.utc(2026, 5, 18, 18, 0, 0));  // → '2026-05-19'
String dhakaDateKey(DateTime now) {
  final dhakaTime = now.toUtc().add(const Duration(hours: 6));
  return DateFormat('yyyy-MM-dd').format(dhakaTime);
}
